qualtran/bloqs/rotations/programmable_ancilla_rotation.py

```diff
@@ -0,0 +1,173 @@
+        eps: precision of the synthesized state.
+    """
+    exponent: SymbolicFloat
+    eps: SymbolicFloat = 1e-11
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature([Register("q", QBit(), side=Side.RIGHT)])
+
+    def build_composite_bloq(self, bb: 'BloqBuilder') -> dict[str, 'SoquetT']:
+        q = bb.allocate(dtype=QBit())
+        q = bb.add(Hadamard(), q=q)
+        q = bb.add(ZPowGate(self.exponent, eps=self.eps), q=q)
+        return {'q': q}
+
+
+@bloq_example
+def _zpow_programmed_ancilla() -> ZPowProgrammedAncilla:
+    zpow_programmed_ancilla = ZPowProgrammedAncilla(np.pi / 4)
+    return zpow_programmed_ancilla
+
+
+@bloq_example
+def _zpow_programmed_ancilla_symb() -> ZPowProgrammedAncilla:
+    t = sympy.Symbol(r"t")
+    zpow_programmed_ancilla_symb = ZPowProgrammedAncilla(t)
+    return zpow_programmed_ancilla_symb
+
+
+@frozen
+class ZPowUsingProgrammedAncilla(Bloq):
+    r"""Single qubit ZPow rotation using resource states.
+
+    This bloq applies a single qubit Z**t rotation only using ZPow resource states and
+    clifford gates. It is designed to exit early as soon as a measurement succeeds.
+
+    Notes:
+        - This bloq uses measurements.
+        - To use this Bloq in costing, use the precise number of rounds that are actually
+          expected during execution. As Qualtran does not support analyzing measurement-based
+          post-selection circuits, the complexity of this Bloq is the worst-case for the
+          chosen number of rounds.
+        - `apply_final_correction` defaults to False, therefore the resource estimates assume
+          that one of the measurements succeeded and the process quit early. To get an exact
+          channel, instead set this parameter to True.
+
+    Args:
+         exponent: The value $t$ to apply $Z**t$ on the input qubit.
+         eps: The precision of the synthesized rotation.
+         n_rounds: The max number of rounds to attempt the rotation.
+         apply_final_correction: Whether to apply an expensive ZPow rotation at
+                 the end to correct the qubit in case all measurements failed.
+
+    References:
+        [Simulating chemistry efficiently on fault-tolerant quantum computers](https://arxiv.org/abs/1204.0567)
+        Jones et. al. 2012. Fig 4.
+    """
+
+    exponent: SymbolicFloat
+    eps: SymbolicFloat = 1e-11
+    n_rounds: SymbolicInt = field(default=2, kw_only=True)
+    apply_final_correction: bool = field(default=False, kw_only=True)
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(q=QBit())
+
+    @classmethod
+    def from_failure_probability(
+        cls,
+        exponent: SymbolicFloat,
+        *,
+        max_fail_probability: SymbolicFloat,
+        eps: SymbolicFloat = 1e-11,
+    ) -> 'ZPowUsingProgrammedAncilla':
+        """Applies the rotation `Rz(angle)` except with some specified failure probability.
```

Comment:
`Rz(angle)` is stale; please update to `ZPow`

Reply:
fixed